src/emulator/extensionsEmulator.ts

```diff
@@ -67,14 +70,32 @@ export class ExtensionsEmulator {
       path.join(os.homedir(), ".cache", "firebase", "extensions");
     const sourceCodePath = path.join(cacheDir, ref);
 
+    // Wait for previous download promise to resolve before we check source validity.
+    // This avoids racing to download the same source multiple times.
+    // Note: The below will not work because it throws the thread to the back of the message queue.
```

Comment:
ooof, sounds like that was a tricky one to debug

Reply:
Thankfully I started with the `if`, or I would have pull my hair out trying to figure out what's going on 😂